File: DSA/DSA_450/Python/06_binary_trees/004_Diameter_Of_Tree.py
"""
Problem: Diameter Of Tree
URL: https://practice.geeksforgeeks.org/problems/diameter-of-binary-tree/1

Problem Statement:
Given a binary tree, find its diameter. Diameter of a tree is the number of nodes on the longest path between any two nodes in the tree.

Sample Input/Output:
Input: 
        1
      /   \
     2     3
    / \   / \
   4   5 6   7

Output: 5
Explanation: Longest path is 4->2->1->3->7 with 5 nodes.
"""

from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def Build_Tree(vals):
    if not vals or vals[0] == -1:
        return None
    root = TreeNode(vals[0])
    q = deque([root])
    i = 1
    while q and i < len(vals):
        node = q.popleft()
        if i < len(vals) and vals[i] != -1:
            node.left = TreeNode(vals[i])
            q.append(node.left)
        i += 1
        if i < len(vals) and vals[i] != -1:
            node.right = TreeNode(vals[i])
            q.append(node.right)
        i += 1
    return root


def Print_Tree(root):
    if not root:
        return
    q = deque([root])
    result = []
    while q:
        node = q.popleft()
        result.append(str(node.val))
        if node.left:
            q.append(node.left)
        if node.right:
            q.append(node.right)
    print(" ".join(result))


class Solution:
    def Diameter_Optimized(self, root):
        """
        Optimized single pass: Calculate height and diameter together
        Time Complexity: O(n) where n is number of nodes
        Space Complexity: O(h) where h is height of tree
        """
        diameter = [0]
        self.Height_Helper(root, diameter)
        return diameter[0]

    def Diameter_Naive(self, root):
        """
        Naive approach: Calculate height at each node
        Time Complexity: O(n^2) worst case
        Space Complexity: O(h) where h is height of tree
        """
        if not root:
            return 0
        left_height = self.Height(root.left)
        right_height = self.Height(root.right)
        diameter_through_root = left_height + right_height + 1
        left_diameter = self.Diameter_Naive(root.left)
        right_diameter = self.Diameter_Naive(root.right)
        return max(diameter_through_root, left_diameter, right_diameter)

    def Height_Helper(self, root, diameter):
        if not root:
            return 0
        left_height = self.Height_Helper(root.left, diameter)
        right_height = self.Height_Helper(root.right, diameter)
        diameter[0] = max(diameter[0], left_height + right_height + 1)
        return 1 + max(left_height, right_height)

    def Height(self, root):
        if not root:
            return 0
        return 1 + max(self.Height(root.left), self.Height(root.right))


def Test_Diameter_Of_Tree():
    solution = Solution()
    
    vals1 = [1, 2, 3, 4, 5, 6, 7]
    root1 = Build_Tree(vals1)
    print("Test 1 - Optimized:", solution.Diameter_Optimized(root1))
    print("Test 1 - Naive:", solution.Diameter_Naive(root1))
    
    vals2 = [1, 2, 3, -1, -1, 4, 5]
    root2 = Build_Tree(vals2)
    print("Test 2 - Optimized:", solution.Diameter_Optimized(root2))
    print("Test 2 - Naive:", solution.Diameter_Naive(root2))
    
    vals3 = [1, 2, -1, 3, -1, 4]
    root3 = Build_Tree(vals3)
    print("Test 3 - Optimized:", solution.Diameter_Optimized(root3))
    print("Test 3 - Naive:", solution.Diameter_Naive(root3))


if __name__ == "__main__":
    Test_Diameter_Of_Tree()

File: DSA/DSA_450/06_binary_trees/004_Diameter_Of_Tree.cpp
/*
Problem: Diameter Of Tree
URL: https://practice.geeksforgeeks.org/problems/diameter-of-binary-tree/1

Problem Statement:
Given a binary tree, find its diameter. Diameter of a tree is the number of nodes on the longest path between any two nodes in the tree.

Sample Input/Output:
Input: 
        1
      /   \
     2     3
    / \   / \
   4   5 6   7

Output: 5
Explanation: Longest path is 4->2->1->3->7 with 5 nodes.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int data;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : data(x), left(NULL), right(NULL) {}
};

TreeNode* Create_Tree(vector<int> vals) {
    if (vals.empty() || vals[0] == -1) return NULL;
    TreeNode* root = new TreeNode(vals[0]);
    queue<TreeNode*> q;
    q.push(root);
    int i = 1;
    while (!q.empty() && i < vals.size()) {
        TreeNode* node = q.front();
        q.pop();
        if (i < vals.size() && vals[i] != -1) {
            node->left = new TreeNode(vals[i]);
            q.push(node->left);
        }
        i++;
        if (i < vals.size() && vals[i] != -1) {
            node->right = new TreeNode(vals[i]);
            q.push(node->right);
        }
        i++;
    }
    return root;
}

class Solution {
public:
    int Diameter_Optimized(TreeNode* root) {
        /*
        Optimized single pass: Calculate height and diameter together
        Time Complexity: O(n) where n is number of nodes
        Space Complexity: O(h) where h is height of tree
        */
        int diameter = 0;
        Height_Helper(root, diameter);
        return diameter;
    }

    int Diameter_Naive(TreeNode* root) {
        /*
        Naive approach: Calculate height at each node
        Time Complexity: O(n^2) worst case
        Space Complexity: O(h) where h is height of tree
        */
        if (!root) return 0;
        int left_height = Height(root->left);
        int right_height = Height(root->right);
        int diameter_through_root = left_height + right_height + 1;
        int left_diameter = Diameter_Naive(root->left);
        int right_diameter = Diameter_Naive(root->right);
        return max({diameter_through_root, left_diameter, right_diameter});
    }

private:
    int Height_Helper(TreeNode* root, int& diameter) {
        if (!root) return 0;
        int left_height = Height_Helper(root->left, diameter);
        int right_height = Height_Helper(root->right, diameter);
        diameter = max(diameter, left_height + right_height + 1);
        return 1 + max(left_height, right_height);
    }

    int Height(TreeNode* root) {
        if (!root) return 0;
        return 1 + max(Height(root->left), Height(root->right));
    }
};

void Test_Diameter_Of_Tree() {
    Solution solution;
    
    vector<int> vals1 = {1, 2, 3, 4, 5, 6, 7};
    TreeNode* root1 = Create_Tree(vals1);
    cout << "Test 1 - Optimized: " << solution.Diameter_Optimized(root1) << endl;
    cout << "Test 1 - Naive: " << solution.Diameter_Naive(root1) << endl;
    
    vector<int> vals2 = {1, 2, 3, -1, -1, 4, 5};
    TreeNode* root2 = Create_Tree(vals2);
    cout << "Test 2 - Optimized: " << solution.Diameter_Optimized(root2) << endl;
    cout << "Test 2 - Naive: " << solution.Diameter_Naive(root2) << endl;
    
    vector<int> vals3 = {1, 2, -1, 3, -1, 4};
    TreeNode* root3 = Create_Tree(vals3);
    cout << "Test 3 - Optimized: " << solution.Diameter_Optimized(root3) << endl;
    cout << "Test 3 - Naive: " << solution.Diameter_Naive(root3) << endl;
}

int main() {
    Test_Diameter_Of_Tree();
    return 0;
}
